### PR TITLE
Add support for YAML config files

### PR DESCRIPTION
**Pull request by austintraver**
* Increase the clarity of the usage documentation
* Add the `requirements.txt` dependency file
* Include the PyYAML package

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,117 +10,140 @@
...
+automatically generated changelog.
+
+This action will:
+
+* **Prepend** each update to the `CHANGELOG.md` file and
+
+* **Commit** the modified `CHANGELOG.md` file.
+
+* (optionally) **Write** the updates as a comment to the pull request of
+  the corresponding release version.
```
Comment (saadmk11) [inline README.md:23]: These are and/or option, So, you can have both or just one of them (Commit or Comment)

File: README.md [hunk 2]
```diff
@@ -10,117 +10,140 @@
...
 **Default Version Number Regex:** This Regex will be checked against a Pull Request title.
-This follows [``SemVer``](https://regex101.com/r/Ly7O1x/3/) (Semantic Versioning) pattern.
-e.g. ``1.0.0``, ``1.0``, ``v1.0.1`` etc.
+This follows [`SemVer`](https://regex101.com/r/Ly7O1x/3/) (Semantic Versioning) pattern.
+e.g. `1.0.0`, `1.0`, `v1.0.1` etc.
 
 For more details on **Semantic Versioning pattern** go to this link: https://regex101.com/r/Ly7O1x/3/
 
-**Note:** [you can provide your own ``regex`` through the ``config`` file](#config-file-usage-optional)
+**Note:** You can provide your own `regex` using [the `config` file](#config-file-usage)
```
Comment (saadmk11) [inline README.md:51]: Reference `(#config-file-usage)` does not match with the new header of that section

File: README.md [hunk 3]
```diff
@@ -10,117 +10,140 @@
...
 
-To **Enable Commenting, Disable Committing, Group changelog items** and some other options
-Look at the [config file docs](#config-file-usage-optional).
+To **Enable Commenting, Disable Committing, Group changelog items** and  
+some other options, see [config file usage (optional)](#config-file-usage-optional).
```
Comment (saadmk11) [inline README.md:54]: Reference `(#config-file-usage-optional)` does not match with the new header of that section

File: README.md [hunk 4]
```diff
@@ -10,117 +10,140 @@
...
+![Changelog CI Status](https://github.com/<username>/<repository>/workflows/Changelog%20CI/badge.svg)
+```
 
 **Output:**
 
 ![Changelog CI Status](https://github.com/saadmk11/changelog-ci/workflows/Changelog%20CI/badge.svg)
 
 
-## Config File Usage (Optional):
+## Configuration
```
Comment (saadmk11) [inline README.md:92]: I think we should keep the ` (Optional):`

File: scripts/changelog-ci.py [hunk 5]
```diff
@@ -1,14 +1,18 @@
...
-# Regex is taken from https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
-# It was modified a little bit to make it a bit less restrictive
-DEFAULT_SEMVER_REGEX = r"v?(0|[1-9]\d*)\.(0|[1-9]\d*)\.?(0|[1-9]\d*)?(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?"
+# The regular expression used to extract semantic versioning is a
+# slightly less restrictive modification of the following regular expression
+# https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
+DEFAULT_SEMVER_REGEX = r"v?(0|[1-9]\d*)\.(0|[1-9]\d*)\.?(0|[1-9]\d*)?(?:-((" \
+                       r"?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[" \
+                       r"1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([" \
+                       r"0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))? "
```
Comment (saadmk11) [inline scripts/changelog-ci.py:15]:
```suggestion
                       r"0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?"
```

File: scripts/changelog-ci.py [hunk 6]
```diff
@@ -65,12 +69,18 @@ def _get_pull_request_title_and_number(event_path):
 
         return title, number
 
-    def _parse_config(self, config_file):
+    def _parse_config(self, filepath: str):
```
Comment (saadmk11) [inline scripts/changelog-ci.py:72]: We are not using annotations anywhere rights now

File: scripts/changelog-ci.py [hunk 7]
```diff
@@ -65,12 +69,18 @@ def _get_pull_request_title_and_number(event_path):
...
         """parse the config file if not provided use default config"""
-        if config_file:
+        if filepath:
             try:
-                with open(config_file, 'r') as config_json:
-                    config = json.load(config_json)
+                file = open(filepath)
```
Comment (saadmk11) [inline scripts/changelog-ci.py:76]: Don't we need to close the file?

File: scripts/changelog-ci.py [hunk 8]
```diff
@@ -1,14 +1,18 @@
 import json
+import yaml
```
Comment (saadmk11) [inline scripts/changelog-ci.py:2]: We should add this with `requests` as it's not a built-in library.

**Review by saadmk11 — CHANGES_REQUESTED** [review]
Thank you for your contribution here :100: 

I have not tested this locally, but the changes look good :100:, I have left some comments regarding minor changes to the documentation. :)

It would be great if you could include a sample `YAML` config in the documentation :) https://github.com/saadmk11/changelog-ci#example-config-file

File: README.md (hunk 1 above)
Comment (austintraver) [reply README.md:23]: Good catch, will resolve this in the next commit.

File: README.md (hunk 4 above)
Comment (austintraver) [reply README.md:92]: I agree that configurations themselves should be optional, but I think users will find this page easier to read if we stick to the common convention. I made this change after consulting `man(1) man`, what I consider to be a somewhat canonical source of truth for writing documentation:

>  A manual page consists of several sections.
> 
> Conventional  section  names  include **NAME**, **SYNOPSIS**, **CONFIGURATION**, **DESCRIPTION**, **OPTIONS** ... **NOTES**, **BUGS**, **EXAMPLE**, **AUTHORS**, and **SEE ALSO**.

Perhaps a compromise: we emphasize that configurations are ***optional*** directly below the `Configuration` heading

File: scripts/changelog-ci.py (hunk 6 above)
Comment (austintraver) [reply scripts/changelog-ci.py:72]: Very well. I will revert this in the next commit.

File: README.md (hunk 4 above)
Comment (saadmk11) [reply README.md:92]: I still feel like it's more straightforward to put it in the header. It will be easy to find and create less confusion :)

Comment (austintraver) [reply README.md:92]: I see where you're coming from. Fixed in https://github.com/saadmk11/changelog-ci/pull/46/commits/a00cd665a8dbfea2bd12149ee7c227181d66caf9

File: scripts/changelog-ci.py (hunk 8 above)
Comment (saadmk11) [reply scripts/changelog-ci.py:2]: I am referring to [PEP 8](https://www.python.org/dev/peps/pep-0008/#imports)

> Imports should be grouped in the following order:
> 
> Standard library imports.
> Related third party imports.
> Local application/library specific imports.
> You should put a blank line between each group of imports.

File: scripts/changelog-ci.py (hunk 7 above)
Comment (saadmk11) [reply scripts/changelog-ci.py:76]: Yes, the garbage collector will eventually close the file, but Isn't it good practice to always explicitly close the file without relying on the garbage collector?

Comment (austintraver) [reply scripts/changelog-ci.py:76]: I agree it is good practice to do so, but my response was answering your first question:

> Don't we need to close the file?

Comment (saadmk11) [reply scripts/changelog-ci.py:76]: Oh, Alright 👍 The PR is almost ready to get merged 💯, I think we should close the opened file here 🙂

File: README.md (hunk 4 above)
Comment (saadmk11) [reply README.md:92]: Looks great, Thanks 💯

File: scripts/changelog-ci.py (hunk 8 above)
Comment (austintraver) [reply scripts/changelog-ci.py]: Ah, thank you for explaining! I understand now. I have pushed an updated commit. https://github.com/saadmk11/changelog-ci/pull/46/commits/92be906b9bdd74d149243d7fb9fe33cc7751707e resolves all issues that violate compliance with PEP 8.

File: scripts/changelog-ci.py (hunk 7 above)
Comment (austintraver) [reply scripts/changelog-ci.py:76]: Go for it! You have write access to this branch

**Comment by saadmk11** [comment]
I don't think we need docker-compose for this action as it only runs one container

File: scripts/changelog-ci.py (hunk 7 above)
Comment (saadmk11) [reply scripts/changelog-ci.py:76]: Sure, I'm merging this to another branch before merging to master as I need to test the YAML config as well, thanks for your great work on this :)

**Review by saadmk11 — APPROVED** [review]
🎉🎉

**Comment by austintraver** [comment]
> I don't think we need docker-compose for this action as it only runs one container

Good point. Compose is truly, as you said, better suited for multi-container applications.

**Pull request merged**